src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs

```diff
@@ -57,159 +65,231 @@ public void Free()
-            if (!this.IsEmpty && _dictionary.TryGetValue(k, out item))
+            get
             {
-                return (item as ArrayBuilder<V>)?.Count ?? 1;
-            }
+                ValueSet valueSet;
+                if (!this.IsEmpty && _dictionary.TryGetValue(k, out valueSet))
+                {
+                    Debug.Assert(valueSet.Count >= 1);
+                    return valueSet.Items;
+                }
 
-            return 0;
+                return ImmutableArray<V>.Empty;
+            }
         }
 
         /// <summary>
-        /// Returns true if one or more items with given key have been added.
+        /// Get a collection of all the keys.
         /// </summary>
-        public bool ContainsKey(K k)
+        public Dictionary<K, ValueSet>.KeyCollection Keys
         {
-            return !this.IsEmpty && _dictionary.ContainsKey(k);
+            get { return this.IsEmpty ? s_emptyDictionary.Keys : _dictionary.Keys; }
         }
 
-        /// <summary>
-        /// Get all values associated with K, in the order they were added.
-        /// Returns empty read-only array if no values were present.
-        /// </summary>
-        public ImmutableArray<V> this[K k]
+        public struct ValueSet : IEnumerable<V>
         {
-            get
+            /// <summary>
+            /// Each value is either a single V or an <see cref="ArrayBuilder{V}"/>.
+            /// Never null.
+            /// </summary>
+            private readonly object _value;
+
+            internal ValueSet(V value)
             {
-                object item;
-                if (!this.IsEmpty && _dictionary.TryGetValue(k, out item))
+                _value = value;
+            }
+
+            internal ValueSet(ArrayBuilder<V> values)
+            {
+                _value = values;
+            }
+
+            internal void Free()
+            {
+                var arrayBuilder = _value as ArrayBuilder<V>;
+                arrayBuilder?.Free();
+            }
+
+            internal V this[int index]
+            {
+                get
                 {
-                    var arrayBuilder = item as ArrayBuilder<V>;
+                    Debug.Assert(this.Count >= 1);
+
+                    var arrayBuilder = _value as ArrayBuilder<V>;
+                    if (arrayBuilder == null)
+                    {
+                        if (index == 0)
+                        {
+                            return (V)_value;
+                        }
+                        else
+                        {
+                            throw new IndexOutOfRangeException();
+                        }
+                    }
+                    else
+                    {
+                        return arrayBuilder[index];
+                    }
+                }
+            }
+
+            internal ImmutableArray<V> Items
+            {
+                get
+                {
+                    Debug.Assert(this.Count >= 1);
```

Comment:
Is it common to fetch same "Items" twice? 
Perhaps it is worth storing the array once it is created. (array can be dropped when set is modified).